contracts/src/systems/transport/contracts/caravan_systems.cairo

```diff
@@ -0,0 +1,133 @@
+            loop {
+                if index == length {
+                    break ();
+                }
+                let entity_id: u128 = *entity_ids[index];
+
+                // assert that they are movable
+                // assert that they have a capacity component
+                let (movable, capacity, position, quantity) = get!(
+                    world, (entity_id), (Movable, Capacity, Position, Quantity)
+                );
+
+                // assert that they are all at the same position when index > 0
+                // if index == 0, then initialize position as the first entity position
+                if index != 0 {
+                    assert(entity_position.x == position.x, 'entities not same position');
+                    assert(entity_position.y == position.y, 'entities not same position');
+                } else {
+                    entity_position = position;
+                }
+
+                // assert that caller is the owner of the entities
+                let owner = get!(world, (entity_id), Owner);
+                assert(caller == owner.address, 'entity is not owned by caller');
+
+                // assert that they are not blocked
+                assert(movable.blocked == false, 'entity is blocked');
+
+                // set entity in the caravan
+                let foreign_key_arr = array![caravan_id.into(), entities_key.into(), index.into()];
+                let foreign_key = poseidon_hash_span(foreign_key_arr.span());
+                set!(world, (ForeignKey { foreign_key, entity_id }));
+
+                // set the entity as blocked so that it cannot be used in another caravan
+                set!(world, (
+                    Movable { 
+                        entity_id, 
+                        sec_per_km: movable.sec_per_km, 
+                        blocked: true,  
+                    })
+                );
+
+                // TODO: add the Caravan component to each entity
+                // so that when we know it's in a caravan
+                let quantity: u128 = quantity.get_value();
+                total_speed += movable.sec_per_km.into() * quantity;
+                total_quantity += quantity;
+                total_capacity += capacity.weight_gram * quantity;
+                index += 1;
+            };
+            // DISCUSS: i created a getAverageSpeed system but
+            // it would mean that we'd have to loop 2x over the entities
+
+            // DISCUSS: i could also create a new CheckAllSamePosition system that checks
+            // if a list of entities are at the same position, but again that would be 
+            // an extra loop
+            let average_speed = total_speed / total_quantity;
+            let average_speed: u16 = average_speed.try_into().unwrap();
+
+            // set the caravan entity
+            set!(world, (
+                    Owner {
+                        entity_id: caravan_id.into(), 
```

Comment:
the `caravan_id.into()` should be cached above for all the following model updates - saves gas